### ISSUE TITLE
Nested object and collection initializers

### ISSUE THREAD

**Issue by AnthonyDGreen** [dotnet/vblang#134]
Today in VB you can initialize an object during construction with the `With` and `From` keywords but this only works for initializing the object itself, not any properties of its members. This is a problem when a member is a read-only property which returns an instanced mutable object because you can't simply assign a new object or collection to the property; the API contract requires you to manipulate the object it gives you.

I propose allowing nested object and collection initializers to solve this problem. This will allow the instantiation and initialization of deeper object graphs in an expression context.

``` VB.NET
Dim order = New Order {
                    .Id = Guid.NewGuid(),
                    .AddressInfo With {
                        .Line1 = "233 S Wacker Dr",
                        .City = "Chicago"
                    }
                    .Items From {
                        New Item With { .ProductId = 1, .Quantity = 1 } }
                    }
                }
```

**Q: Should we support dotted names on the left instead of (or in addition to) the nested With initializer?**

``` VB.NET
Dim order = New Order {
                    .Id = Guid.NewGuid(),
                    .AddressInfo.Line1 = "233 S Wacker Dr",
                    .AddressInfo.City = "Chicago"
                }
```

**Comment by franzalex** [comment]
This will be a nice improvement to object initializers.

Regarding dotted names on the left, I reckon it should simply be left at
level of the nested initializer.
Allowing dotted names could potentially lead to some people shuffling
properties around, making it difficult to realise that the different lines
are actually referencing (and initializing) a single property.

**Comment by AnthonyDGreen** [comment]
@franzalex good point! Especially since it's not clear if each instance re-invokes the property or if it's invoked once and cached for subsequent initializer runs and there's ordering issues. Very good point!
